File: src/ros2_bridge/mappers/std_msgs/uint64_multi_array.rs
//! Typed mapper for `std_msgs/msg/UInt64MultiArray`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn uint64_multi_array_to_bus(msg: ros_env::std_msgs::msg::UInt64MultiArray) -> crate::std_msgs::msg::v1::UInt64MultiArray {
    crate::std_msgs::msg::v1::UInt64MultiArray {
        layout: Some(crate::ros2_bridge::mappers::std_msgs::multi_array_layout::multi_array_layout_to_bus(msg.layout)),
        data: msg.data.into_iter().collect(),
    }
}

pub(crate) fn uint64_multi_array_to_ros(bus: crate::std_msgs::msg::v1::UInt64MultiArray) -> ros_env::std_msgs::msg::UInt64MultiArray {
    ros_env::std_msgs::msg::UInt64MultiArray {
        layout: crate::ros2_bridge::mappers::std_msgs::multi_array_layout::multi_array_layout_to_ros(bus.layout.unwrap_or_default()),
        data: bus.data,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct StdMsgsUInt64MultiArrayMapper;

impl TypedTopicMapper for StdMsgsUInt64MultiArrayMapper {
    type Ros = ros_env::std_msgs::msg::UInt64MultiArray;
    type Bus = crate::std_msgs::msg::v1::UInt64MultiArray;

    fn type_name(&self) -> &'static str {
        "std_msgs/msg/UInt64MultiArray"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(uint64_multi_array_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(uint64_multi_array_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/std_msgs/u_int64_multi_array.py
"""Generated mapper for `std_msgs/msg/UInt64MultiArray`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.multi_array_layout import multi_array_layout_to_bus, multi_array_layout_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.std_msgs.msg.v1 import UInt64MultiArray as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def u_int64_multi_array_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.layout.CopyFrom(multi_array_layout_to_bus(msg.layout))
    bus.data.extend(list(msg.data))
    return bus


def u_int64_multi_array_to_ros(bus):
    from std_msgs.msg import UInt64MultiArray as RosMsg

    out = RosMsg()
    out.layout = multi_array_layout_to_ros(bus.layout)
    out.data = list(bus.data)
    return out


class StdMsgsUInt64MultiArrayMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from std_msgs.msg import UInt64MultiArray as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return u_int64_multi_array_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return u_int64_multi_array_to_ros(bus)
